### PR TITLE
Introduce JsonUtilsNew as documented in #5875

### PR DESCRIPTION
**Pull request by DHowett**
Read the [JsonUtils Spec] for more details.

This pull request introduces the next version of JsonUtils. It is in a
separate file for ease of review and testing.

JsonUtilsNew will be renamed in a subsequent commit that rewrites our
JSON deserializers.

### Implementer's Notes

I went with telescoping exceptions for the key parsing code, because
it's totally possible that you can be five keys deep and encounter a
type error. This lets us encode information about all failures in the
chain instead of just the topmost one.

The original JsonUtilsNew code changed to use `decay` everywhere because
the tests wouldn't compile. We want to treat `GetValue<const guid>` _the
same as_ `GetValue<guid>`, and this lets us do so. `decay` is awesome.

I've been developing this with a shim that redirects `JsonUtils.h` to
`JsonUtilsNew.h`. I am not comfortable deleting the original until we've
moved off of it, and that _will_ be the subject of a followup PR.

## Validation Steps Performed

So many tests.

[JsonUtils Spec]: https://github.com/microsoft/terminal/blob/master/doc/cascadia/Json-Utility-API.md

Refs #2550

### CODE REVIEW

File: src/cascadia/TerminalApp/JsonUtilsNew.h [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+
+Abstract:
+- Helpers for the TerminalApp project
+Author(s):
+- Mike Griese - August 2019
+- Dustin Howett - January 2020
+
+--*/
+#pragma once
+
```
Comment (DHowett) [inline src/cascadia/TerminalApp/JsonUtilsNew.h:16]: nit: fix spacing here

File: src/cascadia/ut_app/JsonUtilsTests.cpp [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+        // Basic string
+        Json::Value stringFirst{ "first" };
+        VERIFY_ARE_EQUAL(JsonTestEnum::First, GetValue<JsonTestEnum>(stringFirst));
+
+        Json::Value stringSecond{ "second" };
+        VERIFY_ARE_EQUAL(JsonTestEnum::Second, GetValue<JsonTestEnum>(stringSecond));
+
+        // Unknown value should produce something?
+        Json::Value stringUnknown{ "unknown" };
+        VERIFY_ARE_EQUAL(JsonTestEnum::First, GetValue<JsonTestEnum>(stringUnknown));
```
Comment (DHowett) [inline src/cascadia/ut_app/JsonUtilsTests.cpp:360]: SOCIALIZE BEHAVIOR?

**Comment by DHowett** [comment]
```
Summary of Non-passing Tests:
    TerminalAppUnitTests::JsonUtilsTests::FlagMapper [Failed]
```

This is expected :)

File: src/cascadia/ut_app/JsonUtilsTests.cpp (hunk 2 above)
Comment (DHowett) [reply src/cascadia/ut_app/JsonUtilsTests.cpp:360]: Do not socialize behavior. Documentation says that it throws.

File: src/cascadia/TerminalApp/JsonUtilsNew.h [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+                    if (found)
+                    {
+                        value |= *found;
+                    }
+                }
+                return value;
+            }
+
+            // If we didn't recognize any flags, the default is _NO FLAGS_
+            return static_cast<T>(0);
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/JsonUtilsNew.h:279]: This should be impossible to hit because of `CanConvert` right?

Comment (carlos-zamora) [reply src/cascadia/TerminalApp/JsonUtilsNew.h:279]: If that's the case, FailFast?

File: src/cascadia/TerminalApp/JsonUtilsNew.h [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+    struct ConversionTrait<til::color>
+    {
+        til::color FromJson(const Json::Value& json)
+        {
+            return ::Microsoft::Console::Utils::ColorFromHexString(Detail::GetStringView(json));
+        }
+
+        bool CanConvert(const Json::Value& json)
+        {
+            return json.isString();
```
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalApp/JsonUtilsNew.h:205]: Yep. We do!

File: src/cascadia/TerminalApp/JsonUtilsNew.h (hunk 3 above)
Comment (DHowett) [reply src/cascadia/TerminalApp/JsonUtilsNew.h:279]: I'd rather not introduce more failfasts if I can help it. I'm 100% on the side of "don't take down the entire application unless there's nothing you can do to recover"

**Comment by DHowett** [comment]
I want some input on how these exceptions should be designed. I don't know if exceptions are the place for localized text...

File: src/cascadia/TerminalApp/JsonUtilsNew.h [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+            }
+            else if (json.isArray())
+            {
+                unsigned int seen{ 0 };
+                T value{};
+                for (const auto& element : json)
+                {
+                    const auto newFlag{ EnumMapper::FromJson(element) };
+                    if (++seen > 1 &&
+                        ((newFlag == AllClear && value != AllClear) ||
```
Comment (DHowett) [inline src/cascadia/TerminalApp/JsonUtilsNew.h:270]: NOTE: I tried `(value|newFlag) == value`, but it makes `"all", "first"` _invalid_. I don't see a compelling reason to make that invalid, whereas for `"none", "first"` I do.

**Comment by miniksa** [comment]
> JsonUtilsNew

image

**Comment by carlos-zamora** [comment]
> I want some input on how these exceptions should be designed. I don't know if exceptions are the place for localized text...

It would be cool if we could display these exceptions as warnings/errors, then localize them like that. But I think that would mean that we localize the displayed warnings/errors, not the exceptions. So you're probably fine?

**Comment by DHowett** [comment]
> It would be cool if we could display these exceptions as warnings/errors

I mean, like, that _is_ the point of having an exception here... not just so that parsing aborts and tells you nothing 😉

**Comment by miniksa** [comment]
We talked about this today. My vote was that we should just go with the localized string with as much data as might be necessary to solve the problem.

**Comment by DHowett** [comment]
I ended up deviating from the discussion anyway, because I cannot guarantee resource loader access during the tests. We'll fix up error messages in post.

**Pull request merged**